runtimebp/cpu.go

```diff
@@ -48,52 +36,18 @@ func NumCPU() (n float64) {
 		return
 	}
 
-	period, err = readNumberFromFile(periodPath, buf)
-	if err != nil {
-		return
+	// CFS quota returns -1 if there is no limit.
+	// Return the default NumCPU for this case to avoid squashing CPU on unlimited quota.
+	if quota < 0 {
+		return float64(runtime.NumCPU())
 	}
 
-	return quota / period
-}
-
-// On some situations the quota file will be -1,
-// and we should use this one instead.
-//
-// Those situations include:
-//
-// - Very old version of docker
-//
-// - In k8s only request is set for cpu, not limit
-func numCPUSharesFallback() (n float64) {
-	const (
-		sharesPath  = "/sys/fs/cgroup/cpu/cpu.shares"
-		denominator = 1024
-	)
-
-	var err error
-	defer func() {
-		if err != nil || n <= 0 {
-			// Fallback and log to stderr.
-			fmt.Fprintf(
-				os.Stderr,
-				"NumCPU: falling back to use runtime.NumCPU(): %v, %v\n",
-				n,
-				err,
-			)
-			n = float64(runtime.NumCPU())
-		}
-	}()
-
-	// Big enough buffer to read the number in the file wholly into memory.
-	buf := make([]byte, 1024)
-	var shares float64
-
-	shares, err = readNumberFromFile(sharesPath, buf)
+	period, err = readNumberFromFile(periodPath, buf)
 	if err != nil {
 		return
```

Comment:
with the removal of `defer` this will return 0. if you only want to remove the fallback to shares you should just change line 38 in the original code.

as to whether we want to remove the fallback to shares, that's still up to debate.

Reply:
Good catch, I'll fix it.